examples/shape.rs

```diff
@@ -0,0 +1,344 @@
+}
+
+fn shape_into(
+    font_index: usize,
+    text: &str,
+    script: u32,
+    lang: u32,
+    loaded_fonts: &[LoadedFont],
+    results: &mut Vec<Shaped>,
+) -> Result<()> {
+    let first_pass = loaded_fonts[font_index].shape_text(text, font_index, script, lang)?;
+
+    let mut item_iter = first_pass.into_iter();
+    let mut fallback_run = String::new();
+    while let Some(item) = item_iter.next() {
+        match &item.info {
+            ResolvedGlyph::Resolved(_) => results.push(item),
+            ResolvedGlyph::Unresolved(raw) => {
+                // There was no glyph in that font, so we'll need to shape
+                // using a fallback.  Let's collect together any potential
+                // run of unresolved entries first
+                for &c in &raw.unicodes {
+                    fallback_run.push(c);
+                }
+
+                while let Some(item) = item_iter.next() {
+                    match &item.info {
+                        ResolvedGlyph::Unresolved(raw) => {
+                            for &c in &raw.unicodes {
+                                fallback_run.push(c);
+                            }
+                        }
+                        ResolvedGlyph::Resolved(_) => {
+                            shape_into(
+                                font_index + 1,
+                                &fallback_run,
+                                script,
+                                lang,
+                                loaded_fonts,
+                                results,
+                            )?;
+                            fallback_run.clear();
+                            results.push(item);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    if !fallback_run.is_empty() {
+        shape_into(
+            font_index + 1,
+            &fallback_run,
+            script,
+            lang,
+            loaded_fonts,
+            results,
+        )?;
+    }
+
+    Ok(())
+}
+
+struct LoadedFont {
+    cmap_subtable: CmapSubtable<'static>,
+    gpos_cache: Option<LayoutCache<GPOS>>,
+    gsub_cache: LayoutCache<GSUB>,
+    gdef_table: Option<GDEFTable>,
+    hmtx: HmtxTable<'static>,
+    hhea: HheaTable,
+    num_glyphs: u16,
+
+    // Must be last: this keeps the 'static items alive
+    _scope: ReadScopeOwned,
+}
+
+impl LoadedFont {
+    fn load_font<P: AsRef<std::path::Path>>(path: P) -> Result<Self> {
+        let data = std::fs::read(path.as_ref())?;
+        let owned_scope = ReadScopeOwned::new(ReadScope::new(&data));
+
+        // This unsafe block and transmute are present so that we can
+        // extend the lifetime of the OpenTypeFile that we produce here.
+        // That in turn allows us to store all of these derived items
+        // into a struct and manage their lifetimes together.
+        let file: OpenTypeFile<'static> =
+            unsafe { transmute(owned_scope.scope().read::<OpenTypeFile>()?) };
```

Comment:
I don't like this transmute: it would be great if there were a clear and easy way to end up with an owned instance so that the lifetimes are easier to deal with for the consuming code.